### PR TITLE
nixos/borgbackup: add an option to ignore warnings

### PR DESCRIPTION
**Pull request by symphorien**
borgbackup exit code 1 means warning, not error. For example if a file is modified during backup, borg create exits with code 1. It may be desirable to count that as success.

See also https://github.com/NixOS/nixpkgs/pull/201969 which does the same but by patching source code instead.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/backup/borgbackup.nix [hunk 1]
```diff
@@ -33,13 +33,26 @@ let
     }
     trap on_exit EXIT
 
+    borgWrapper () {
+      set +e
+      borg "$@"
+      local result=$?
+      set -e
```
Comment (bjornfor) [inline nixos/modules/services/backup/borgbackup.nix:40]: I think `set` will affect the caller. Is there any downside to running it in a subshell?

File: nixos/modules/services/backup/borgbackup.nix [hunk 2]
```diff
@@ -33,13 +33,26 @@ let
...
+      if [[ -z "${toString cfg.failOnWarnings}" ]] && [[ "$result" == 1 ]]; then
+        echo "ignoring warning return value 1";
+        return 0;
+      else
+        return "$result";
```
Comment (bjornfor) [inline nixos/modules/services/backup/borgbackup.nix:45]: Unneeded trailing semicolons.

Comment (symphorien) [reply nixos/modules/services/backup/borgbackup.nix]: removed

File: nixos/modules/services/backup/borgbackup.nix (hunk 1 above)
Comment (symphorien) [reply nixos/modules/services/backup/borgbackup.nix:40]: what do you mean affects the caller? I'm only restoring set -e, so this is expected.
If you meant that set +e line 37 could affect concurrent execution, apparently it does not:
```bash
set -e

fun () {
    set +e
    sleep 2
    false
    echo "false ignored in fun"
    set -e
}

fun2 () {
    sleep 1
    echo "here we know that fun has run set +e"
    false
    echo "we never get here"
}

fun | fun2
```
```
:!bash -x foo.sh
+ set -e
+ fun
+ set +e
+ sleep 2
+ fun2
+ sleep 1
+ echo 'here we know that fun has run set +e'
+ false
here we know that fun has run set +e
+ false
+ echo 'false ignored in fun'

le shell a retourné 1
```

About using a subshell, I think I can't define variables in subshells which are also available in the outer shell.

**Comment by Scrumplex** [comment]
I think it might be easier to configure `SuccessExitStatus=` for the systemd service instead.

**Comment by symphorien** [comment]
> I think it might be easier to configure SuccessExitStatus= for the systemd service instead.

that would only work if the service consisted only of one borg invocation. With set -e, if borg create returns a warning and we exit, borg prune/compact are not run.

File: nixos/modules/services/backup/borgbackup.nix (hunk 1 above)
Comment (bjornfor) [reply nixos/modules/services/backup/borgbackup.nix:40]: I was thinking about the interaction between the module shell code and the pre/post hook shell code provided by the user, which I think run in the same process. I haven't looked closely, but I'm always cautious about these stateful operations.

Yes, variables defined in sub-shells stay in the sub-shell process, but I was thinking about something like .... hm, you're right, that's difficult to get right with subshells/without parent variables. But how about

```
# either errexit state works with this code, so we don't have to set it at all
#set -e
#set +e

echo "pre-borg"
(exit 1) && rc=$? || rc=$?   # simulate borg exit code
echo "post-borg rc=$rc"
```

Comment (symphorien) [reply nixos/modules/services/backup/borgbackup.nix:40]: thanks that works. I changed the PR in favor of this idiom.

**Review by bjornfor — APPROVED** [review]

**Pull request merged**